### PR TITLE
Display both unread and starred message counts in navigation drawer

### PR DESCRIPTION
**Pull request by andrewfg**
**Issue:**

In v5.8 the navigation drawer would only display the `unread` message count in the fragment badges.

**Solution:**

This PR displays both `unread` and `flagged` message counts in the fragment badges (see screen shot below)

**Note:**

**The PR is not yet complete!!**
So far it only has the front end UI code, and it is still missing the back end code to fetch the `flagged` message count from the mail server.

image

### CODE REVIEW

**Comment by cketti** [comment]
Thanks for taking the time to create a pull request :heart: 

Regarding the development process:
- While the pull request is not ready yet, please mark it as a draft. 
- For questions around development, please use the [discussions](https://github.com/k9mail/k-9/discussions) section here on GitHub, not the forum. (I'm also happy to answer questions in our [Matrix](https://matrix.to/#/#k9mail:matrix.org) chat channel. But let's try to keep generic questions to GitHub, so others can benefit from the answers, too.)

Regarding the feature:
- I don't want this to be the default. I believe only displaying the unread count makes the user interface far less cluttered and thus more beginner-friendly.
- I don't particularly like adding settings either. But this seems like functionality enough people seem to want (back). So making it an optional feature is probably the way to go.

It's fine working on the feature first, then adding a setting later.

**Comment by andrewfg** [comment]
> While the pull request is not ready yet, please mark it as a draft.

The PR labels box is read only, so I just added the text "DRAFT" to the title.

> I don't want this to be the default

Ok.

**Comment by cketti** [comment]
The link to convert an existing pull request to a draft is not easy to find. It's in the "reviewers" section.

image

**Comment by andrewfg** [comment]
Ok, marked as DRAFT now.

I think that I will need help with fetching the `flagged message count` from the backend. In prior versions, there was code to fetch it, but it seems that in v5.x that code was deleted. And due to the refactoring of the project structure from v4.x to v5.x I cannot figure out where the respective code got to..

**Comment by cketti** [comment]
> I think that I will need help with fetching the `flagged message count` from the backend.

You'll have to add another subquery for the number of starred messages. It's basically the same as this, only with checking the `messages.flagged` column instead of `messages.read`:

https://github.com/k9mail/k-9/blob/03fb2b953840a667f452c8a8d65ddb15f84b8348/app/storage/src/main/java/com/fsck/k9/storage/messages/RetrieveFolderOperations.kt#L72-L76

Let me know if you're still having trouble making the backend changes. I'm happy to provide a patch.

We sometimes use the term "flagged messages", sometimes "starred messages". I think "starred" is clearer and we should use that in new code.

**Comment by andrewfg** [comment]
> It's basically the same as this, only with checking the messages.flagged column instead of messages.read

Thanks for the info. I will have another attempt.

> I think "starred" is clearer and we should use that in new code.

Understood. Actually I started using "starred" and then refactored all the names to "flagged" in order to be more backwards compatible! But I can easily re-re-factor it again.

**Comment by andrewfg** [comment]
Hi @cketti, I just pushed a commit with the following changes..

- Refactored `flaggedCount` back to `starredCount` again
- Added a second SQL sub query to calculate `starredCount`

I have tested this, and I can confirm that it **works perfectly** for the Navigation Drawer **"Folder View"**. But unfortunately I was not able to do the coding for the "Account View" and the "Unified Inbox View" as shown "TODO" in the links below. Basically I don't understand enough about the Kotlin Flow concept to be able write this code: => so I would appreciate it if you could give me any tips :)

[K9Drawer.kt](https://github.com/k9mail/k-9/blob/63e8ec15f8024f100f3ffd73da6619cad29f7586/app/ui/legacy/src/main/java/com/fsck/k9/ui/K9Drawer.kt#:~:text=//%20TODO%3A%20fetch%20unified,val%20starredCount%20%3D%2099)
image

[AccountsViewModel.kt](https://github.com/k9mail/k-9/blob/63e8ec15f8024f100f3ffd73da6619cad29f7586/app/ui/legacy/src/main/java/com/fsck/k9/ui/account/AccountsViewModel.kt#:~:text=//%20TODO%3A%20fetch%20starred,index%5D%2C%20unreadCount%2C%20starredCount)
image

**Comment by andrewfg** [comment]
The original intent of this PR was to display the following..
1. Starred and unread message counts in **folder view** navigation drawer items
2. Starred and unread message counts in **account view** navigation drawer items
3. Starred and unread message counts in **unified inbox** navigation drawer item

I have coded and tested item 1. above, but I lack the coding skills for items 2. and 3. -- to be specfic..
- How to fetch from the back end the starred message counts for **account view** navigation drawer items
- How to fetch from the back end the starred and unread message counts for **unified inbox** navigation drawer item

So I have decided to split this PR into two parts -- namely item 1. resp. items 2. & 3. With today's commit the functionality of item 1. is assured and tested. And the functionality of item 2. & 2. is suppressed (there are code stubs `TODO` but nothing will display).

I am therefore converting this part 1. PR from `DRAFT` to `READY FOR REVIEW`

**Comment by andrewfg** [comment]
@cketti IMHO this PR is now fully ready with all functionality and fully tested .. (at least it is fully ready for review)

**Comment by andrewfg** [comment]
PS I also added a new setting for "Show Starred Message Count" with a default value of `FALSE`

**Comment by andrewfg** [comment]
IMHO this should solve https://github.com/k9mail/k-9/issues/5570

**Comment by andrewfg** [comment]
image

**Review by cketti — CHANGES_REQUESTED** [review]
You seem to be passing the composite data type for unread count and starred count (`IntArray` in your case) to a lot of places that really only need the unread count. This results in a lot more changes than are necessary. Please undo these changes and only pass the unread count along.

File: app/core/src/main/java/com/fsck/k9/controller/UnreadMessageCountProvider.kt [hunk 1]
```diff
@@ -12,8 +12,8 @@ import com.fsck.k9.search.getAccounts
 import timber.log.Timber
 
 interface UnreadMessageCountProvider {
-    fun getUnreadMessageCount(account: Account): Int
-    fun getUnreadMessageCount(searchAccount: SearchAccount): Int
+    fun getUnreadStarredCount(account: Account): IntArray
+    fun getUnreadStarredCount(searchAccount: SearchAccount): IntArray
 }
```
[the remark this answers is not in the document]
Comment (andrewfg) [reply app/core/src/main/java/com/fsck/k9/controller/UnreadMessageCountProvider.kt:17]: Ok. Done.

File: app/core/src/main/java/com/fsck/k9/mailstore/LocalStore.java [hunk 2]
```diff
@@ -1275,6 +1275,40 @@ public Integer doDbWork(SQLiteDatabase db) throws WrappedException, MessagingExc
...
 
+    private int getStarredCount(LocalSearch search) throws MessagingException {
+        StringBuilder whereBuilder = new StringBuilder();
+        List<String> queryArgs = new ArrayList<>();
+        SqlQueryBuilder.buildWhereClause(account, search.getConditions(), whereBuilder, queryArgs);
+
+        String where = whereBuilder.toString();
+        final String[] selectionArgs = queryArgs.toArray(new String[queryArgs.size()]);
+
+        final String sqlQuery = "SELECT SUM(flagged<>0) " +
```
[the remark this answers is not in the document]
Comment (andrewfg) [reply app/core/src/main/java/com/fsck/k9/mailstore/LocalStore.java:1286]: Ok. Done.

File: app/k9mail/src/main/java/com/fsck/k9/widget/unread/UnreadWidgetData.kt [hunk 3]
```diff
@@ -5,6 +5,6 @@ import android.content.Intent
 data class UnreadWidgetData(
     val configuration: UnreadWidgetConfiguration,
     val title: String,
-    val unreadCount: Int,
+    val unreadStarredCount: IntArray,
```
[the remark this answers is not in the document]
Comment (andrewfg) [reply app/k9mail/src/main/java/com/fsck/k9/widget/unread/UnreadWidgetData.kt:8]: Ok. Done.

File: app/storage/src/main/java/com/fsck/k9/storage/messages/RetrieveFolderOperations.kt [hunk 4]
```diff
@@ -74,6 +74,12 @@ internal class RetrieveFolderOperations(private val lockableDatabase: LockableDa
                     WHERE messages.folder_id = folders.id 
                       AND messages.empty = 0 AND messages.deleted = 0 
                       AND (messages.read = 0 OR folders.id = ?)
+                ), (
+                    SELECT COUNT(messages.id) 
+                    FROM messages 
+                    WHERE messages.folder_id = folders.id 
+                      AND messages.empty = 0 AND messages.deleted = 0 
+                      AND (messages.flagged <> 0 OR folders.id = ?)
```
[the remark this answers is not in the document]
Comment (andrewfg) [reply app/storage/src/main/java/com/fsck/k9/storage/messages/RetrieveFolderOperations.kt:82]: Ok. Done.

File: app/ui/legacy/src/main/res/values/strings.xml [hunk 5]
```diff
@@ -833,6 +833,7 @@ Please submit bug reports, contribute new features and ask questions at
     <string name="volume_navigation_list">In list views</string>
 
     <string name="show_unified_inbox_title">Show Unified Inbox</string>
+    <string name="show_starred_count_title">Show Starred Count</string>
```
[the remark this answers is not in the document]
Comment (andrewfg) [reply app/ui/legacy/src/main/res/values/strings.xml:836]: Ok. Done.

File: app/core/src/main/java/com/fsck/k9/K9.kt [hunk 6]
```diff
@@ -307,6 +310,7 @@ object K9 : EarlyInit {
         isUseVolumeKeysForNavigation = storage.getBoolean("useVolumeKeysForNavigation", false)
         isUseVolumeKeysForListNavigation = storage.getBoolean("useVolumeKeysForListNavigation", false)
         isShowUnifiedInbox = storage.getBoolean("showUnifiedInbox", true)
+        isShowStarredCount = storage.getBoolean("showStarredCount", false)
```
[the remark this answers is not in the document]
Comment (andrewfg) [reply app/core/src/main/java/com/fsck/k9/K9.kt:313]: Ok. Done. But please check the `version` count.

File: app/ui/legacy/src/main/java/com/fsck/k9/ui/K9Drawer.kt [hunk 7]
```diff
@@ -305,6 +328,15 @@ class K9Drawer(private val parent: MessageList, savedInstanceState: Bundle?) : K
                 selectedColorInt = selectedBackgroundColor
                 textColor = selectedTextColor
                 isSelected = unifiedInboxSelected
+
+                var unreadStarredCount = messagingController.getUnreadStarredCount(unifiedInbox)
```
[the remark this answers is not in the document]
Comment (andrewfg) [reply app/ui/legacy/src/main/java/com/fsck/k9/ui/K9Drawer.kt:332]: Ok. I will comment this out and mark as TODO

File: app/core/src/main/java/com/fsck/k9/mailstore/LocalFolder.java [hunk 8]
```diff
@@ -316,6 +320,36 @@ public Integer doDbWork(final SQLiteDatabase db) throws WrappedException {
...
+            open();
+        }
+
+        try {
+            return this.localStore.getDatabase().execute(false, new DbCallback<Integer>() {
+                @Override
+                public Integer doDbWork(final SQLiteDatabase db) throws WrappedException {
+                    int unreadMessageCount = 0;
+                    Cursor cursor = db.query("messages", new String[] { "COUNT(id)" },
+                            "folder_id = ? AND empty = 0 AND deleted = 0 AND flagged<>0",
```
[the remark this answers is not in the document]
Comment (andrewfg) [reply app/core/src/main/java/com/fsck/k9/mailstore/LocalFolder.java]: Ok. Done.

File: app/ui/legacy/src/main/java/com/fsck/k9/ui/K9Drawer.kt [hunk 9]
```diff
@@ -162,6 +164,26 @@ class K9Drawer(private val parent: MessageList, savedInstanceState: Bundle?) : K
...
     }
 
+    /**
+     * Format the unread and starred counts for display in the fragment Badge box
+     */
+    private fun formatBadgeText(unreadStarredCount: IntArray): String {
+        var badgeText = ""
+        if (K9.isShowStarredCount) {
+            if (unreadStarredCount[0] > 0) {
+                badgeText = "● " + unreadStarredCount[0].toString()
```
[the remark this answers is not in the document]
Comment (andrewfg) [reply app/ui/legacy/src/main/java/com/fsck/k9/ui/K9Drawer.kt]: Ok. Done.

File: app/ui/legacy/src/main/java/com/fsck/k9/ui/K9Drawer.kt [hunk 10]
```diff
@@ -162,6 +164,26 @@ class K9Drawer(private val parent: MessageList, savedInstanceState: Bundle?) : K
...
+            }
+            if (unreadStarredCount[1] > 0) {
+                badgeText = badgeText + "   ★ " + unreadStarredCount[1].toString()
```
[the remark this answers is not in the document]
Comment (andrewfg) [reply app/ui/legacy/src/main/java/com/fsck/k9/ui/K9Drawer.kt]: Ok. Done.

**Comment by andrewfg** [comment]
@cketti many thanks for the review. I think I have addressed all your issues, so it is ready to review again IMHO.

Also I would like to suggest that if the user has more than one account, then the navigation drawer should **open by default in the account view** (below), whereas if the user has just one account it should open in folder view. => What do you think?

image

**Comment by cketti** [comment]
You can use `./gradlew ktlintFormat` to fix Kotlin code style violations.

**Comment by andrewfg** [comment]
> You can use ./gradlew ktlintFormat to fix Kotlin code style violations.

Thanks for the tip. Generally I use Eclipse IDE which does auto formatting on save. So Android Studio seems rather retrogade in not having that feature.

**Comment by andrewfg** [comment]
Phew! All CI tests passed - finally..

File: app/core/src/main/java/com/fsck/k9/preferences/GeneralSettingsDescriptions.java [hunk 11]
```diff
@@ -283,6 +283,9 @@
         s.put("showRecentChanges", Settings.versions(
                 new V(73, new BooleanSetting(true))
         ));
+        s.put("showStarredCount", Settings.versions(
+                new V(73, new BooleanSetting(false))
```
Comment (cketti) [inline app/core/src/main/java/com/fsck/k9/preferences/GeneralSettingsDescriptions.java:287]: Regarding the version, see:

https://github.com/k9mail/k-9/blob/f6e9948c2ad3b4056592fb77015b0da0f3907cfb/app/core/src/main/java/com/fsck/k9/preferences/GeneralSettingsDescriptions.java#L47-L50

File: app/ui/legacy/src/main/java/com/fsck/k9/ui/K9Drawer.kt [hunk 12]
```diff
@@ -83,6 +84,9 @@ class K9Drawer(private val parent: MessageList, savedInstanceState: Bundle?) : K
     private var openedAccountUuid: String? = null
     private var openedFolderId: Long? = null
 
+    // TODO: unifiedInbox needed to get its message counts from back end
+    // private val unifiedInbox: SearchAccount = SearchAccount.createUnifiedInboxAccount()
```
Comment (cketti) [inline app/ui/legacy/src/main/java/com/fsck/k9/ui/K9Drawer.kt:88]: Please don't leave any commented out code in the pull request. If a feature isn't ready, it doesn't go into the code base.

**Review by cketti — CHANGES_REQUESTED** [review]

File: app/core/src/main/java/com/fsck/k9/preferences/GeneralSettingsDescriptions.java (hunk 11 above)
Comment (andrewfg) [reply app/core/src/main/java/com/fsck/k9/preferences/GeneralSettingsDescriptions.java:287]: Ok. Done. (Bumped to version 75)

File: app/core/src/main/java/com/fsck/k9/mailstore/FolderMapper.kt [hunk 13]
```diff
@@ -21,7 +22,7 @@ interface FolderDetailsAccessor {
     val pushClass: FolderClass
     val visibleLimit: Int
     val moreMessages: MoreMessages
-    val messageCount: Int
+    val messageCounts: MessageCounts
```
[the remark this answers is not in the document]
Comment (andrewfg) [reply app/core/src/main/java/com/fsck/k9/mailstore/FolderMapper.kt:25]: Probably need to do DisplayAccount as well as DisplayFolder for consistency. Or ??

File: app/ui/legacy/src/main/java/com/fsck/k9/ui/K9Drawer.kt (hunk 12 above)
Comment (andrewfg) [reply app/ui/legacy/src/main/java/com/fsck/k9/ui/K9Drawer.kt:88]: Ok. Done.

File: app/core/src/main/java/com/fsck/k9/mailstore/FolderMapper.kt (hunk 13 above)
[the remark this answers is not in the document]
Comment (andrewfg) [reply app/core/src/main/java/com/fsck/k9/mailstore/FolderMapper.kt:25]: ^
Cannot do above due to interference with `displayAccountFlow: Flow<List<DisplayAccount>>`

[the remark this answers is not in the document]
Comment (andrewfg) [reply app/core/src/main/java/com/fsck/k9/mailstore/FolderMapper.kt]: FolderMapper and DisplayFolder done.

**Review by cketti — APPROVED** [review]

**Comment by andrewfg** [comment]
@cketti I like the changes you have made. Obviously some reflect your personal coding style preferences, but others are very clever. Please excuse the difficult birth of this PR; although I have decades of coding experience on many platforms, in many languages, and in many IDEs, this was my first on Android / Kotlin / Android Studio. So in fact a quite steep learning process. Thank you for your assistance in this.

**Comment by cketti** [comment]
Thanks for contributing to K-9 Mail :heart: 

> Please excuse the difficult birth of this PR

Forget about it. This went a lot smoother than the average pull request of comparable complexity :+1:

**Pull request merged**